File: backend/importClaims.js
import fs from 'node:fs/promises'
import path from 'node:path'
import { parse } from 'csv-parse/sync'
import 'dotenv/config'
import { buildMemberDocuments, normalizeClaim } from './claimMapper.js'
import { closeMongo, connectMongo, getMongoConfig } from './db.js'
import { predictClaim } from '../shared/predictionEngine.js'

const defaultCsvPath = '/Users/user/Downloads/EDI_834_837_20 members(837_Claims).csv'
const csvPath = process.argv[2] || process.env.CSV_PATH || defaultCsvPath

function chunk(items, size) {
  const chunks = []
  for (let index = 0; index < items.length; index += size) {
    chunks.push(items.slice(index, index + size))
  }
  return chunks
}

async function ensureIndexes(db) {
  const claims = db.collection('claims')
  const members = db.collection('members')
  const predictions = db.collection('claim_predictions')

  await claims.createIndex({ claimId: 1 }, { unique: true })
  await claims.createIndex({ number: 1 }, { unique: true })
  await claims.createIndex({ memberId: 1, dos: -1 })
  await claims.createIndex({ payer: 1, dos: -1 })
  await claims.createIndex({ billingProvider: 1, dos: -1 })
  await claims.createIndex({ patient: 'text', memberId: 'text', number: 'text' })

  await members.createIndex({ memberId: 1 }, { unique: true })
  await members.createIndex({ patient: 'text', memberId: 'text' })

  await predictions.createIndex({ claimId: 1 }, { unique: true })
  await predictions.createIndex({ number: 1 }, { unique: true })
  await predictions.createIndex({ memberId: 1, dos: -1 })
  await predictions.createIndex({ 'prediction.risks.overall.score': -1 })
}

async function bulkUpsert(collection, documents, keyField) {
  let modified = 0
  let upserted = 0

  for (const batch of chunk(documents, 500)) {
    if (!batch.length) continue

    const result = await collection.bulkWrite(
      batch.map((document) => ({
        replaceOne: {
          filter: { [keyField]: document[keyField] },
          replacement: document,
          upsert: true,
        },
      })),
      { ordered: false },
    )

    modified += result.modifiedCount
    upserted += result.upsertedCount
  }

  return { modified, upserted }
}

async function main() {
  const resolvedCsvPath = path.resolve(csvPath)
  const csv = await fs.readFile(resolvedCsvPath, 'utf8')
  const rows = parse(csv, {
    columns: true,
    skip_empty_lines: true,
    trim: true,
  })

  const claims = rows
    .map(normalizeClaim)
    .filter((claim) => claim.claimId && claim.memberId)
    .sort((a, b) => b.dos.localeCompare(a.dos))
  const members = buildMemberDocuments(claims)

  const db = await connectMongo()
  await ensureIndexes(db)

  const claimResult = await bulkUpsert(db.collection('claims'), claims, 'claimId')
  const memberResult = await bulkUpsert(db.collection('members'), members, 'memberId')
  const predictionDocs = claims.map((claim) => ({
    claimId: claim.claimId,
    number: claim.number,
    memberId: claim.memberId,
    patient: claim.patient,
    payer: claim.payer,
    billingProvider: claim.billingProvider,
    dos: claim.dos,
    prediction: predictClaim(claim, claims),
    predictedAt: new Date(),
  }))
  const predictionResult = await bulkUpsert(db.collection('claim_predictions'), predictionDocs, 'claimId')

  if (process.env.SYNC_DELETE === 'true') {
    await db.collection('claims').deleteMany({ claimId: { $nin: claims.map((claim) => claim.claimId) } })
    await db.collection('members').deleteMany({ memberId: { $nin: members.map((member) => member.memberId) } })
  }

  const totals = await db.collection('claims').aggregate([
    {
      $group: {
        _id: null,
        totalClaims: { $sum: 1 },
        totalCharges: { $sum: '$totalCharge' },
        totalAllowed: { $sum: '$allowed' },
        totalPaid: { $sum: '$paid' },
        totalPatientResp: { $sum: '$patientResp' },
      },
    },
  ]).toArray()

  console.log(JSON.stringify({
    database: getMongoConfig().dbName,
    csvPath: resolvedCsvPath,
    parsedRows: rows.length,
    importedClaims: claims.length,
    importedMembers: members.length,
    claimResult,
    memberResult,
    predictionResult,
    totals: totals[0] || {},
  }, null, 2))
}

main()
  .catch((error) => {
    console.error(error.message)
    process.exitCode = 1
  })
  .finally(async () => {
    await closeMongo()
  })

File: backend/import_claims.py
"""Import the 837 CSV into MongoDB.

Usage: python3 -m backend.import_claims /path/to/claims.csv
"""

import csv
import json
import os
import sys
from pathlib import Path

from pymongo import ReplaceOne

from .claim_mapper import build_member_documents, normalize_claim
from .db import close_mongo, connect_mongo, get_mongo_config


def read_claims(path):
    with path.open(newline="", encoding="utf-8-sig") as csv_file:
        rows = csv.DictReader(csv_file)
        return [claim for claim in (normalize_claim(row) for row in rows) if claim.get("claimId") and claim.get("memberId")]


def upsert(collection, documents, key):
    if not documents:
        return {"matched": 0, "upserted": 0}
    result = collection.bulk_write(
        [ReplaceOne({key: document[key]}, document, upsert=True) for document in documents],
        ordered=False,
    )
    return {"matched": result.matched_count, "upserted": result.upserted_count}


def main():
    source = sys.argv[1] if len(sys.argv) > 1 else os.getenv("CSV_PATH")
    if not source:
        raise SystemExit("Provide a CSV path: python3 -m backend.import_claims /path/to/claims.csv")
    csv_path = Path(source).expanduser().resolve()
    claims = read_claims(csv_path)
    members = build_member_documents(claims)
    db = connect_mongo()
    if getattr(db, "is_fallback", False):
        raise SystemExit("MongoDB is unavailable. Set MONGODB_URI before importing the CSV.")

    db.claims.create_index("claimId", unique=True)
    db.claims.create_index([("memberId", 1), ("diagnosisCode", 1), ("dos", -1)])
    db.members.create_index("memberId", unique=True)
    results = {
        "database": get_mongo_config()["dbName"],
        "csvPath": str(csv_path),
        "claims": len(claims),
        "members": len(members),
        "claimWrite": upsert(db.claims, claims, "claimId"),
        "memberWrite": upsert(db.members, members, "memberId"),
    }
    print(json.dumps(results, indent=2))


if __name__ == "__main__":
    try:
        main()
    finally:
        close_mongo()
